lib/constructs/vpc-endpoint-lb-target.ts

```diff
@@ -0,0 +1,170 @@
+    }
+
+    // This is the custom reasource responsible for building the JSON data that can be used
+    // to specify target information for the Load Balancer. This has to be a custom resource
+    // because the information that we need (the IP addresses of the ENIs associated with the
+    // VPC Endpoint) is not available as an attribute of the CloudFormation resource. In fact,
+    // because it requires multiple API calls (ec2:DescribeVpcEndpoints
+    // and ec2:DescribeNetworkInterfaces) we cannot use the nicer `AwsCustomResource` construct
+    // and instead must write our own handler with the Provider framework.
+    const handler = new nodejs.NodejsFunction(this, "VpcEndpointHandler", {
+      runtime: lambda.Runtime.NODEJS_16_X,
+      entry: "lib/custom-resources/endpoint-ips.ts",
+      handler: "onEvent",
+      initialPolicy: [
+        new iam.PolicyStatement({
+          effect: iam.Effect.ALLOW,
+          actions: ["ec2:DescribeVpcEndpoints", "ec2:DescribeNetworkInterfaces"],
+          resources: ["*"],
+        }),
+      ],
+    });
+    const vpcEndpointIpProvider = new cr.Provider(this, "VpcEndpointIps", {
+      onEventHandler: handler,
+    });
+    const customResource = new cdk.CustomResource(this, "ApiGatewayEndpointIps", {
+      serviceToken: vpcEndpointIpProvider.serviceToken,
+      properties: {
+        VpcEndpointId: endpoint.vpcEndpointId,
+        Port: targetGroupProps.port ?? 443,
+      },
+    });
+
+    // This is the key within the `Data` field of the custom resource.
+    this.targetJson = customResource.getAtt("Targets");
+  }
+}
+
+/**
+ * Allows using a VPC Endpoint as the target for an Application Load Balancer
+ */
+export class VpcEndpointApplicationTargetGroup
+  extends VpcEndpointBaseTargetGroup<elbv2.ApplicationTargetGroup>
+  implements elbv2.IApplicationTargetGroup
+{
+  constructor(scope: Construct, id: string, props: VpcEndpointApplicationTargetGroupProps) {
+    super(scope, id, props);
+    this.targetGroup = new elbv2.ApplicationTargetGroup(this, "TargetGroup", props);
+    const targetGroupCfn = this.targetGroup.node.defaultChild as elbv2.CfnTargetGroup;
+    targetGroupCfn.addPropertyOverride("Targets", this.targetJson);
+    targetGroupCfn.addPropertyOverride("TargetType", "ip");
+  }
+
+  registerListener(listener: elbv2.IApplicationListener, associatingConstruct?: IConstruct | undefined): void {
+    return this.targetGroup.registerListener(listener, associatingConstruct);
+  }
+
+  registerConnectable(connectable: ec2.IConnectable, portRange?: ec2.Port | undefined): void {
+    return this.targetGroup.registerConnectable(connectable, portRange);
+  }
+
+  addTarget(...targets: elbv2.IApplicationLoadBalancerTarget[]): void {
+    throw new Error("Additional targets are not supported for VpcEndpointApplicationTargetGroup");
+  }
+}
+
+/**
+ * Allows using a VPC Endpoint as the target for a Network Load Balancer.
+ */
+export class VpcEndpointNetworkTargetGroup
```

Comment:
Really like the use of OO design to cleanly support both types of ELBs (perhaps showing my bias as a Java guy at heart)